### PR TITLE
WartRemover and Coursier for comm

### PR DESCRIPTION
**Pull request by rabbitonweb**
This PR does two things:

1. adds WartRemover
This is initial configuration of WartRemover. I've noticed that many
classes should not be case classes or if they are, they should be
marked as final. Added WartRemover to do the work for me.

This configuration will be extended in the future, while we keep on
improving towards more FP code base

2. add Coursier 
It adds Coursier for parallel dependency fetching & solving some well known Ivy issues

### CODE REVIEW

**Comment by rabbitonweb** [comment]
@kirkwood this is ready for review

File: comm/src/main/scala/coop/rchain/comm/comm.scala [hunk 1]
```diff
@@ -40,7 +43,7 @@ case class PeerNode(id: NodeIdentifier, endpoint: Endpoint) {
     s"rnode://$sKey@${endpoint.host}:${endpoint.udpPort}"
 
   def withUdpSocket(udp: java.net.InetSocketAddress): PeerNode =
-    PeerNode(id, endpoint.withUdp(udp))
+    new PeerNode(id, endpoint.withUdp(udp))
```
Comment (kirkwood) [inline comm/src/main/scala/coop/rchain/comm/comm.scala]: Wait . . . what?

Comment (rabbitonweb) [reply comm/src/main/scala/coop/rchain/comm/comm.scala]: yeah, this should be reverted back, forgot about it, let me elaborate on this a bit

`PeerNode` is case class but really should not be as you have some class other class that extends it. This breaks the equals method and you will have some unexpected errors on runtime if you leave it as it is. 
I tried to make `PeerNode` not a case class but then realized that I don't know enough of domain to touch it just yet. So I reverted that back (making `PeerNode` case class one more time) leaving a `TODO/FIX-ME` and a comment.

This is something I hope we could discuss really quickly and one of the reasons I like static analysis like that as it finds those kinds of errror - thus WartRemover

Regardless of `PeerNode` being case class or not, for now the above should be reverted back to previous version

**Comment by rabbitonweb** [comment]
@kirkwood https://github.com/rchain/rchain/pull/230#discussion_r166339291

discussion on `PeerNode` needed but I will revert that `new` now

**Comment by henrytill** [comment]
Using wartremover as an sbt plugin doesn't work unfortunately for those of us who use IntelliJ.

Instead, if you add the wartremover traversers directly to `scalacOptions`, then it works with both sbt and IntelliJ.  You can do something like:
```scala
lazy val wartremoverOptions = Seq(
  "Any",
  "ArrayEquals",
  "AsInstanceOf",
  "DefaultArguments",
  "EitherProjectionPartial",
  "Enumeration",
  "Equals",
  "ExplicitImplicitTypes",
  "FinalCaseClass",
  "FinalVal",
  "ImplicitConversion",
  "IsInstanceOf",
  "JavaConversions",
  "LeakingSealed",
  "MutableDataStructures",
  "NonUnitStatements",
  "Nothing",
  "Null",
  "Option2Iterable",
  "OptionPartial",
  "Overloading",
  "PublicInference",
  "Product",
  "Return",
  "Serializable",
  "StringPlusAny",
  "Throw",
  "ToString",
  "TraversableOps",
  "TryPartial",
  "Var",
  "While"
).map((s: String) => "-P:wartremover:traverser:org.wartremover.warts." + s)

lazy val nonConsoleOptions =
  Set("-Ywarn-unused-import", "-Xfatal-warnings") ++ wartremoverOptions.toSet

// ...
  scalacOptions ++= compilerOptions ++ wartremoverOptions,
  scalacOptions in (Compile, console) ~= { _.filterNot(nonConsoleOptions) },
  scalacOptions in (Test, console) := (scalacOptions in (Compile, console)).value,
// ...
```

File: comm/src/main/scala/coop/rchain/comm/comm.scala (hunk 1 above)
Comment (kirkwood) [reply comm/src/main/scala/coop/rchain/comm/comm.scala]: Got it; thanks.

**Comment by rabbitonweb** [comment]
@henrytill I will modify it tomorrow and ask you for a comment once its there, thx

**Comment by rabbitonweb** [comment]
@henrytill wait, one thing though, why would you filter out it during compile 
` scalacOptions in (Compile, console) ~= { _.filterNot(nonConsoleOptions) },` ?

**Comment by henrytill** [comment]
This has to do with scoping in sbt.  

You are not filtering it out during the `compile` task (in the `Compile` dependency configuration).  You are filtering it out during the `console` task (also in the `Compile` dependency configuration).  

See _Scoping by the configuration axis_:
https://www.scala-sbt.org/1.0/docs/Scopes.html

Using wartremover and some other `scalacOption`s at the REPL  is a recipe for pain IME.  I believe the `sbt-wartremover` plugin takes care of this for you, though I could be wrong.

**Comment by rabbitonweb** [comment]
@henrytill could you have a look at last commit regarding making IDE happy? thx!

File: comm/project/plugins.sbt [hunk 2]
```diff
@@ -3,3 +3,8 @@ addSbtPlugin("se.marcuslonnberg" % "sbt-docker" % "1.5.0")
 addSbtPlugin("com.geirsson" % "sbt-scalafmt" % "1.4.0")
 
 addSbtPlugin("org.scoverage" % "sbt-scoverage" % "1.5.1")
+
+addSbtPlugin("org.wartremover" % "sbt-wartremover" % "2.2.1")
```
Comment (henrytill) [inline comm/project/plugins.sbt]: For the new pan-IDE configuration, remove this.

File: comm/build.sbt [hunk 3]
```diff
@@ -39,6 +39,7 @@ libraryDependencies ++= Seq(
   "ch.qos.logback" % "logback-classic" % "1.2.3"
 )
 
+
 addCompilerPlugin(
   "org.scalamacros" % "paradise" % "2.1.0" cross CrossVersion.full
 )
```
Comment (henrytill) [inline comm/build.sbt]: For the new pan-IDE wartremover configuration, add

```scala
addCompilerPlugin("org.wartremover" %% "wartremover" % "2.3.1")
```

File: comm/build.sbt [hunk 4]
```diff
@@ -60,8 +61,15 @@ lazy val commonOptions = Seq(
...
+// ignore autogenerated sources from protobuf
+// this does not work perfectly yet, see build.sh for details
+wartremoverExcluded := ((sourceManaged.value) ** "*.scala").get
+
+lazy val wartremoverOptions = Seq(
+  "FinalCaseClass"
+).map((s: String) => "-P:wartremover:traverser:org.wartremover.warts." + s)
 
-scalacOptions ++= commonOptions
+scalacOptions ++= commonOptions ++ wartremoverOptions
```
Comment (henrytill) [inline comm/build.sbt]: 👍

**Comment by rabbitonweb** [comment]
@henrytill rdy

**Comment by henrytill** [comment]
Looking good!

**Pull request merged**